### DISCUSSION TITLE
Display characters instead of numbers inside the LegacyMainCircle (Piece)

### DISCUSSION THREAD

**Discussion by 0xdeadbeer** [ppy/osu#18850] [Q&A] [answered]
I'll start this discussion by saying that it has nothing to do with the original osu! lazer (just working on a private osu! lazer version). 
For the last couple of days I've been trying to display normal text instead of a 0-9 digit number inside the circles.. unfortunately none of what I tried/tested worked. Figured out the location of where the number is "added" to the circle, which is inside `/osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyMainCirclePiece.cs`. 

*notice the bold line
<pre>
protected override void LoadComplete()
        {
            base.LoadComplete();

            accentColour.BindValueChanged(colour => CircleSprite.Colour = LegacyColourCompatibility.DisallowZeroAlpha(colour.NewValue), true);
            if (hasNumber)
                <b>indexInCurrentCombo.BindValueChanged(index => hitCircleText.Text = (index.NewValue + 1).ToString(), true);</b>

            if (drawableObject != null)
            {
                drawableObject.ApplyCustomUpdateState += updateStateTransforms;
                updateStateTransforms(drawableObject, drawableObject.State.Value);
            }
        }
</pre>

Tried to switch that line with something like `hitCircleText.Text = @"nice"`.. which did compile, it did run, but it left me with something like this:
![osu_2022-06-26_15-03-39](https://user-images.githubusercontent.com/64986162/175815435-0b68458e-6e19-42ef-a259-e06529054e09.png)

Also have noticed the Font property of the hitCircleText (since that `OsuFont.Numeric.With(size: 40)` seemed like a problem to me), and I've switched it to `OsuFont.Default.With(size: 40)`... and it did absolutely nothing.

<pre>
if (hasNumber)
            {
                OverlayLayer.Add(hitCircleText = new SkinnableSpriteText(new OsuSkinComponent(OsuSkinComponents.HitCircleText), _ => new OsuSpriteText
                {
                    <b>Font = OsuFont.Numeric.With(size: 40),</b>
                    UseFullGlyphHeight = false,
                }, confineMode: ConfineMode.NoScaling)
                {
                    Anchor = Anchor.Centre,
                    Origin = Anchor.Centre,
                });
            }
</pre>

I'm very confused as to why this is happening and how I could fix it.. any kind of help is highly appreciated!

**Comment by bdach** [comment, marked as answer]
i'm not sure i should be responding in the context of you creating a "private osu!lazer version" because that sounds like nothing good, but i'll bite.

what you're changing there ("the bold line") is the default fallback implementation of the `HitCircleText` component. go read how `SkinnableDrawable` works, and then go read this:

https://github.com/ppy/osu/blob/4bc26dbb487241e2bbae73751dbe9e93a4e427da/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs#L109-L117

figuring out the rest of why it's not working is left as an exercise to the reader

**Reply by 0xdeadbeer** [reply]
Why would creating a private osu! lazer sound bad? The code is under MIT license why would there be any problems? Either way the code will be public once I finish creating my first demo so..

Thanks for the hint! Will report back once I'll fix the problem :smile:

**Reply by bdach** [reply]
the code may be MIT, but the associated resources are not and the associated trademarks are not, as per https://github.com/ppy/osu#licence

just something to keep in mind. and if you have suggestions of new features or other improvements in this "private version" of yours, then a nicer way of going about it would be to just ask us here.

**Reply by 0xdeadbeer** [reply]
Got that! Thanks for telling me.
Anyway I fixed the issue, everything just makes sense now thanks for your help!
And no I don't see any reason to share my ideas with you all as I'm very sure they're not inside your.. scope.
But if I'll get any that may fit the original osu!lazer I'll definitely share it rest assured :smile:
